Index: src/root/shared/new_types.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/root/shared/new_types.rs b/src/root/shared/new_types.rs
new file mode 100644
--- /dev/null	(date 1725206804893)
+++ b/src/root/shared/new_types.rs	(date 1725206804893)
@@ -0,0 +1,46 @@
+use std::collections::HashMap;
+use either::Either;
+use itertools::Itertools;
+use crate::root::parser::location::Location;
+
+pub struct GlobalTable {
+    type_templates: HashMap<String, UserTypeTemplate>,
+    cached_types: HashMap<TypeID, >
+}
+
+#[derive(Eq, PartialEq, Hash)]
+pub enum TypeID {
+    User { id: usize, parameters: Vec<TypeID>, array_count: usize },
+    Builtin { id: usize }
+}
+
+
+#[derive(Eq, PartialEq)]
+pub struct TraitID(usize);
+
+pub struct UserTypeTemplate {
+    location: Location,
+    parameters: Vec<Vec<TraitID>>,
+    attributes: Vec<(String, Either<TypeID, usize>)> // (Name, Either<Type, index of parameter>) 
+}
+
+impl UserTypeTemplate {
+    pub fn fill<'a>(&'a self, parameters: &'a [TypeID]) -> UserType {
+        UserType {
+            location: &self.location,
+            size: 0,
+            attributes: self.attributes.iter().map(|(name, t)| {
+                match t {
+                    Either::Left(t) => (0, name.as_str(), t),
+                    Either::Right(i) => (0, name.as_str(), &parameters[*i])
+                }
+            }).collect_vec()
+        }
+    }
+}
+
+pub struct UserType<'a> {
+    location: &'a Location,
+    size: usize,
+    attributes: Vec<(usize, &'a str, &'a TypeID)>
+}
\ No newline at end of file
Index: src/root/shared/mod.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>pub(crate) mod common;\npub mod types;\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/root/shared/mod.rs b/src/root/shared/mod.rs
--- a/src/root/shared/mod.rs	(revision 49f4fdbbcb63757a1d88ee6db34e3d0b9d001f3e)
+++ b/src/root/shared/mod.rs	(date 1725206807305)
@@ -1,2 +1,3 @@
 pub(crate) mod common;
 pub mod types;
+mod new_types;
